# Exercici 1

<b> Agafa un text en anglès que vulguis i calcula'n la freqüència de les paraules. </b>

Hem escollit el següent article:

<p style="font-size:18px;">What is data science?</p>

*Data science combines math and statistics, specialized programming, advanced analytics, artificial intelligence (AI), and machine learning with specific subject matter expertise to uncover actionable insights hidden in an organization’s data. These insights can be used to guide decision making and strategic planning.*

*The accelerating volume of data sources, and subsequently data, has made data science is one of the fastest growing field across every industry. As a result, it is no surprise that the role of the data scientist was dubbed the “sexiest job of the 21st century” by Harvard Business Review (link resides outside of IBM). Organizations are increasingly reliant on them to interpret data and provide actionable recommendations to improve business outcomes.*

*The data science lifecycle involves various roles, tools, and processes, which enables analysts to glean actionable insights. Typically, a data science project undergoes the following stages:*

*Data ingestion: The lifecycle begins with the data collection--both raw structured and unstructured data from all relevant sources using a variety of methods. These methods can include manual entry, web scraping, and real-time streaming data from systems and devices. Data sources can include structured data, such as customer data, along with unstructured data like log files, video, audio, pictures, the Internet of Things (IoT), social media, and more.*

*Data storage and data processing: Since data can have different formats and structures, companies need to consider different storage systems based on the type of data that needs to be captured. Data management teams help to set standards around data storage and structure, which facilitate workflows around analytics, machine learning and deep learning models. This stage includes cleaning data, deduplicating, transforming and combining the data using ETL (extract, transform, load) jobs or other data integration technologies. This data preparation is essential for promoting data quality before loading into a data warehouse, data lake, or other repository.*

*Data analysis: Here, data scientists conduct an exploratory data analysis to examine biases, patterns, ranges, and distributions of values within the data. This data analytics exploration drives hypothesis generation for a/b testing. It also allows analysts to determine the data’s relevance for use within modeling efforts for predictive analytics, machine learning, and/or deep learning. Depending on a model’s accuracy, organizations can become reliant on these insights for business decision making, allowing them to drive more scalability.*

*Communicate: Finally, insights are presented as reports and other data visualizations that make the insights—and their impact on business—easier for business analysts and other decision-makers to understand. A data science programming language such as R or Python includes components for generating visualizations; alternately, data scientists can use dedicated visualization tools.*

Enllaç: https://www.ibm.com/topics/data-science

En primer lloc assignem el text a una variable. Alternativament podríem llegir el text des de l'enllaç, mitjançant llibreries com 'requests' per obtenir el contingut de la pàgina i 'BeautifulSoup' per a parsejar el contingut HTML i extreure'n el text.

In [1]:
text = """What is data science?

Data science combines math and statistics, specialized programming, advanced analytics, artificial intelligence (AI), and machine learning with specific subject matter expertise to uncover actionable insights hidden in an organization’s data. These insights can be used to guide decision making and strategic planning.

The accelerating volume of data sources, and subsequently data, has made data science is one of the fastest growing field across every industry. As a result, it is no surprise that the role of the data scientist was dubbed the “sexiest job of the 21st century” by Harvard Business Review (link resides outside of IBM). Organizations are increasingly reliant on them to interpret data and provide actionable recommendations to improve business outcomes.

The data science lifecycle involves various roles, tools, and processes, which enables analysts to glean actionable insights. Typically, a data science project undergoes the following stages:

Data ingestion: The lifecycle begins with the data collection--both raw structured and unstructured data from all relevant sources using a variety of methods. These methods can include manual entry, web scraping, and real-time streaming data from systems and devices. Data sources can include structured data, such as customer data, along with unstructured data like log files, video, audio, pictures, the Internet of Things (IoT), social media, and more.

Data storage and data processing: Since data can have different formats and structures, companies need to consider different storage systems based on the type of data that needs to be captured. Data management teams help to set standards around data storage and structure, which facilitate workflows around analytics, machine learning and deep learning models. This stage includes cleaning data, deduplicating, transforming and combining the data using ETL (extract, transform, load) jobs or other data integration technologies. This data preparation is essential for promoting data quality before loading into a data warehouse, data lake, or other repository.

Data analysis: Here, data scientists conduct an exploratory data analysis to examine biases, patterns, ranges, and distributions of values within the data. This data analytics exploration drives hypothesis generation for a/b testing. It also allows analysts to determine the data’s relevance for use within modeling efforts for predictive analytics, machine learning, and/or deep learning. Depending on a model’s accuracy, organizations can become reliant on these insights for business decision making, allowing them to drive more scalability.

Communicate: Finally, insights are presented as reports and other data visualizations that make the insights—and their impact on business—easier for business analysts and other decision-makers to understand. A data science programming language such as R or Python includes components for generating visualizations; alternately, data scientists can use dedicated visualization tools."""

Ara farem un anàlisi de la freqüència de paraules del text, tal i com demana l'exercici.

In [2]:
# Importem les llibreries necessàries
import nltk
from nltk.tokenize import word_tokenize

# Fem una tokenització
tokens = word_tokenize(text)

# Fem una conversió a minúscules
tokens = [word.lower() for word in tokens]

# Fem el càlcul de freqüència de paraules
word_freq = nltk.FreqDist(tokens)

# Mostrem les paraules i les seves freqüències
for word, freq in word_freq.items():
    print(word, ":", freq)


what : 1
is : 4
data : 40
science : 6
? : 1
combines : 1
math : 1
and : 18
statistics : 1
, : 40
specialized : 1
programming : 2
advanced : 1
analytics : 4
artificial : 1
intelligence : 1
( : 4
ai : 1
) : 4
machine : 3
learning : 5
with : 3
specific : 1
subject : 1
matter : 1
expertise : 1
to : 12
uncover : 1
actionable : 3
insights : 5
hidden : 1
in : 1
an : 2
organization : 1
’ : 3
s : 3
. : 19
these : 3
can : 6
be : 2
used : 1
guide : 1
decision : 2
making : 2
strategic : 1
planning : 1
the : 16
accelerating : 1
volume : 1
of : 9
sources : 3
subsequently : 1
has : 1
made : 1
one : 1
fastest : 1
growing : 1
field : 1
across : 1
every : 1
industry : 1
as : 4
a : 6
result : 1
it : 2
no : 1
surprise : 1
that : 3
role : 1
scientist : 1
was : 1
dubbed : 1
“ : 1
sexiest : 1
job : 1
21st : 1
century : 1
” : 1
by : 1
harvard : 1
business : 4
review : 1
link : 1
resides : 1
outside : 1
ibm : 1
organizations : 2
are : 2
increasingly : 1
reliant : 2
on : 5
them : 2
interpret : 1
provide : 1
rec

Anem a explicar detalladament que hem fet. Primer, s'ha importat NLTK i el mòdul word_tokenize per a la tokenització del text. A continuació, s'ha aplicat la tokenització al text, dividint-lo en paraules individuals o "tokens". Després, aquestes paraules s'han convertit totes a minúscules per estandarditzar-les i evitar duplicacions per diferències en majúscules o minúscules. Finalment, s'ha calculat la freqüència de cada paraula utilitzant FreqDist de NLTK, que proporciona un recompte de quants cops apareix cada paraula en el text. Finalment, s'han mostrat les paraules i les seves freqüències associades.

També podem crear un dataframe de Pandas per mostrar la llista de paraules de manera més organitzada i llegible:

In [3]:
import pandas as pd

# Convertim a DataFrame
df = pd.DataFrame(word_freq.items(), columns=['Paraula', 'Freqüència'])

# Ordenem les dades per freqüència en ordre descendent
df = df.sort_values(by='Freqüència', ascending=False)

# Mostrem els primers resultats del dataframe
df.head(10)

Paraula  Freqüència
2       data          40
9          ,          40
36         .          19
7        and          18
46       the          16
26        to          12
49        of           9
194      for           7
3    science           6
62         a           6

# Exercici 2

<b> Treu les stopwords i realitza stemming al teu conjunt de dades.</b>

In [4]:
# Importem les llibreries
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

# Fem una tokenització
tokens = word_tokenize(text)

# Eliminem les stopwords
stop_words = set(stopwords.words('english')) # Indiquem que l'idioma és l'anglès
tokens_without_stopwords = [word for word in tokens if word.lower() not in stop_words]

# Mostrem el resultat
print(tokens_without_stopwords)

['data', 'science', '?', 'Data', 'science', 'combines', 'math', 'statistics', ',', 'specialized', 'programming', ',', 'advanced', 'analytics', ',', 'artificial', 'intelligence', '(', 'AI', ')', ',', 'machine', 'learning', 'specific', 'subject', 'matter', 'expertise', 'uncover', 'actionable', 'insights', 'hidden', 'organization', '’', 'data', '.', 'insights', 'used', 'guide', 'decision', 'making', 'strategic', 'planning', '.', 'accelerating', 'volume', 'data', 'sources', ',', 'subsequently', 'data', ',', 'made', 'data', 'science', 'one', 'fastest', 'growing', 'field', 'across', 'every', 'industry', '.', 'result', ',', 'surprise', 'role', 'data', 'scientist', 'dubbed', '“', 'sexiest', 'job', '21st', 'century', '”', 'Harvard', 'Business', 'Review', '(', 'link', 'resides', 'outside', 'IBM', ')', '.', 'Organizations', 'increasingly', 'reliant', 'interpret', 'data', 'provide', 'actionable', 'recommendations', 'improve', 'business', 'outcomes', '.', 'data', 'science', 'lifecycle', 'involves',

Ara farem una neteja i una lematització per tal de millorar la qualitat i la consistència del nostre conjunt de dades. La neteja eliminarà tot el soroll innecessari, com ara signes de puntuació i caràcters especials, assegurant que el text quedi lliure de distorsions i més enfocat per a l'anàlisi. La lematització, per la seva banda, normalitzarà les paraules a la seva forma bàsica i significativa, tractant les diferents formes d'una mateixa paraula com un únic ítem. Això és crucial per a una interpretació més precisa en anàlisis lingüístiques i semàntiques, especialment en tasques de classificació de textos, modelització de temes i mineria de text. Ambdues tècniques, neteja i lematització, són fonamentals per obtenir un conjunt de dades net, coherent i òptim per a qualsevol tipus d'anàlisi de processament de llenguatge natural.

Comencem eliminant els signes de puntuació i càracters no desitjats:

In [5]:
# Eliminem els signes de puntuació i nombres
tokens_without_punctuation = [word for word in tokens_without_stopwords if word.isalpha()]

# Mostrem el resultat 
print(tokens_without_punctuation)

['data', 'science', 'Data', 'science', 'combines', 'math', 'statistics', 'specialized', 'programming', 'advanced', 'analytics', 'artificial', 'intelligence', 'AI', 'machine', 'learning', 'specific', 'subject', 'matter', 'expertise', 'uncover', 'actionable', 'insights', 'hidden', 'organization', 'data', 'insights', 'used', 'guide', 'decision', 'making', 'strategic', 'planning', 'accelerating', 'volume', 'data', 'sources', 'subsequently', 'data', 'made', 'data', 'science', 'one', 'fastest', 'growing', 'field', 'across', 'every', 'industry', 'result', 'surprise', 'role', 'data', 'scientist', 'dubbed', 'sexiest', 'job', 'century', 'Harvard', 'Business', 'Review', 'link', 'resides', 'outside', 'IBM', 'Organizations', 'increasingly', 'reliant', 'interpret', 'data', 'provide', 'actionable', 'recommendations', 'improve', 'business', 'outcomes', 'data', 'science', 'lifecycle', 'involves', 'various', 'roles', 'tools', 'processes', 'enables', 'analysts', 'glean', 'actionable', 'insights', 'Typica

Ara fem una lematització:

In [6]:
# Importem el mòdul/llibreria
from nltk.stem import WordNetLemmatizer

# Inicialitzem el lemmatizer
lemmatizer = WordNetLemmatizer()

# Apliquem la lematització
lemmatized_tokens = [lemmatizer.lemmatize(word) for word in tokens_without_punctuation]

# Mostrem el resultat
print(lemmatized_tokens)

['data', 'science', 'Data', 'science', 'combine', 'math', 'statistic', 'specialized', 'programming', 'advanced', 'analytics', 'artificial', 'intelligence', 'AI', 'machine', 'learning', 'specific', 'subject', 'matter', 'expertise', 'uncover', 'actionable', 'insight', 'hidden', 'organization', 'data', 'insight', 'used', 'guide', 'decision', 'making', 'strategic', 'planning', 'accelerating', 'volume', 'data', 'source', 'subsequently', 'data', 'made', 'data', 'science', 'one', 'fastest', 'growing', 'field', 'across', 'every', 'industry', 'result', 'surprise', 'role', 'data', 'scientist', 'dubbed', 'sexiest', 'job', 'century', 'Harvard', 'Business', 'Review', 'link', 'resides', 'outside', 'IBM', 'Organizations', 'increasingly', 'reliant', 'interpret', 'data', 'provide', 'actionable', 'recommendation', 'improve', 'business', 'outcome', 'data', 'science', 'lifecycle', 'involves', 'various', 'role', 'tool', 'process', 'enables', 'analyst', 'glean', 'actionable', 'insight', 'Typically', 'data',

# Exercici 3

<b> Realitza un anàlisi de sentiment al teu conjunt de dades. </b>

In [7]:
# Importem la llibreria/mòdul
from nltk.sentiment import SentimentIntensityAnalyzer

# Inicialitzem l'analitzador de sentiments
sia = SentimentIntensityAnalyzer()

# Obtenim les puntuacions de sentiments
sentiment_scores = sia.polarity_scores(text)

# Mostrem el resultat
print(sentiment_scores)

{'neg': 0.005, 'neu': 0.925, 'pos': 0.07, 'compound': 0.9719}


El resultat de l'anàlisi de sentiments de l'article sobre ciència de dades de la pàgina corporativa d'IBM indica una predominança de tonalitat neutra amb una tendència lleugerament positiva. 

La puntuació negativa és extremadament baixa (neg: 0.005), mostrant una quasi absència de sentiments negatius, el que és esperable en un contingut corporatiu informatiu. La majoria del contingut és classificat com neutre (neu: 0.925), reflectint l'enfocament objectiu i factual que és l'esperat en publicacions corporatives i articles tècnics. Tot i això, hi ha una presència notable de sentiments positius (pos: 0.07), que podria reflectir un to optimista sobre el tema de la ciència de dades. La puntuació composta (compound: 0.9719) reafirma aquesta tonalitat globalment positiva, tot i que lleugera, la qual pot ser el resultat de l'expressió d'entusiasme o confiança en les capacitats i impacte de la ciència de dades segons la perspectiva d'IBM. 

Aquest tipus de tonalitat és comú en materials corporatius que busquen destacar els punts forts i les innovacions de l'empresa de manera subtil però efectiva. En resum, un to afirmatiu en un article corporatiu serveix no només per informar, sinó també per influir positivament en la percepció del lector sobre el tema, en aquest cas, la ciència de dades, i potencialment sobre la pròpia empresa, IBM. Aquest tipus de to és sovint utilitzat per posicionar una empresa com a líder de pensament o a l'avantguarda d'un camp específic.